File: app/backend/impl/AutoEvaluationHandlerImpl.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package backend.impl;

import akka.actor.ActorSystem;
import backend.models.AutoEvaluationConfig;
import backend.models.Exam;
import backend.models.Grade;
import backend.models.GradeEvaluation;
import backend.models.GradeScale;
import backend.models.User;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import org.joda.time.DateTime;
import play.Logger;
import scala.concurrent.duration.Duration;

public class AutoEvaluationHandlerImpl implements AutoEvaluationHandler {
  private final EmailComposer composer;

  private final ActorSystem actor;

  private static final Logger.ALogger logger = Logger.of(AutoEvaluationHandlerImpl.class);

  @Inject
  public AutoEvaluationHandlerImpl(EmailComposer composer, ActorSystem actor) {
    this.composer = composer;
    this.actor = actor;
  }

  @Override
  public void autoEvaluate(Exam exam) {
    AutoEvaluationConfig config = exam.getAutoEvaluationConfig();
    if (config != null) {
      // Grade automatically
      process(exam);
      if (config.getReleaseType() == AutoEvaluationConfig.ReleaseType.IMMEDIATE) {
        // Notify student immediately
        exam.setAutoEvaluationNotified(DateTime.now());
        exam.update();
        User student = exam.getCreator();
        actor
          .scheduler()
          .scheduleOnce(
            Duration.create(5, TimeUnit.SECONDS),
            () -> composer.composeInspectionReady(student, null, exam),
            actor.dispatcher()
          );
        logger.debug("Mail sent about automatic evaluation to {}", student.getEmail());
      }
    }
  }

  private void process(Exam exam) {
    Grade grade = getGradeBasedOnScore(exam);
    if (grade != null) {
      exam.setGrade(grade);
      exam.setGradedTime(DateTime.now());
      exam.setCreditType(exam.getExamType());
      // NOTE: do not set graded by person here, one who makes a record will get the honor
      if (!exam.getExamLanguages().isEmpty()) {
        exam.setAnswerLanguage(exam.getExamLanguages().get(0).getCode());
      } else {
        throw new RuntimeException("No exam language found!");
      }
      exam.setState(Exam.State.GRADED);
      exam.update();
    }
  }

  private Optional<GradeScale> resolveScale(Exam exam) {
    if (exam.getGradeScale() != null) {
      return Optional.of(exam.getGradeScale());
    }
    if (exam.getCourse() != null && exam.getCourse().getGradeScale() != null) {
      return Optional.of(exam.getCourse().getGradeScale());
    }
    if (exam.getParent() != null && exam.getParent().getGradeScale() != null) {
      return Optional.of(exam.getParent().getGradeScale());
    } else if (exam.getParent() != null && exam.getParent().getCourse() != null) {
      GradeScale scale = exam.getParent().getCourse().getGradeScale();
      return scale == null ? Optional.empty() : Optional.of(scale);
    }
    return Optional.empty();
  }

  private Grade getGradeBasedOnScore(Exam exam) {
    Double totalScore = exam.getTotalScore();
    Double maxScore = exam.getMaxScore();
    Double percentage = maxScore == 0 ? 0 : totalScore * 100 / maxScore;
    List<GradeEvaluation> gradeEvaluations = new ArrayList<>(exam.getAutoEvaluationConfig().getGradeEvaluations());
    gradeEvaluations.sort(Comparator.comparingInt(GradeEvaluation::getPercentage));
    Grade grade = null;
    Iterator<GradeEvaluation> it = gradeEvaluations.iterator();
    GradeEvaluation prev = null;
    while (it.hasNext()) {
      GradeEvaluation ge = it.next();
      int threshold = 0;
      if (ge.getPercentage() > percentage) {
        grade = prev == null ? ge.getGrade() : prev.getGrade();
        threshold = prev == null ? ge.getPercentage() : prev.getPercentage();
      } else if (!it.hasNext()) {
        // Highest possible grade
        grade = ge.getGrade();
        threshold = ge.getPercentage();
      }
      if (grade != null) {
        logger.info(
          "Automatically grading exam #{}, {}/{} points ({}%) graded as {} using percentage threshold {}",
          exam.getId(),
          totalScore,
          maxScore,
          percentage,
          grade.getName(),
          threshold
        );
        break;
      }
      prev = ge;
    }
    Optional<GradeScale> scale = resolveScale(exam);
    if (scale.isEmpty() || !scale.get().getGrades().contains(grade)) {
      throw new RuntimeException("Grade in auto evaluation configuration not found in exam grade scale!");
    }
    return grade;
  }
}

File: app/services/exam/AutoEvaluationHandlerImpl.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package services.exam

import database.EbeanQueryExtensions
import models.assessment.AutoEvaluationReleaseType
import models.exam.ExamState
import models.exam.{Exam, GradeScale}
import org.joda.time.DateTime
import play.api.Logging
import services.mail.EmailComposer

import javax.inject.Inject
import scala.concurrent.duration.*
import scala.jdk.CollectionConverters.*

class AutoEvaluationHandlerImpl @Inject (
    private val composer: EmailComposer
) extends AutoEvaluationHandler
    with EbeanQueryExtensions
    with Logging:

  override def autoEvaluate(exam: Exam): Unit =
    val config = exam.autoEvaluationConfig
    if Option(config).nonEmpty then
      // Grade automatically
      process(exam)
      if config.releaseType == AutoEvaluationReleaseType.IMMEDIATE then
        // Notify student immediately
        exam.autoEvaluationNotified = DateTime.now
        exam.update()
        val student = exam.creator
        composer.scheduleEmail(5.seconds) { composer.composeInspectionReady(student, None, exam) }
        logger.debug(s"Scheduled email about automatic evaluation to ${student.email}")

  private def process(exam: Exam): Unit =
    getGradeBasedOnScore(exam) match
      case Left(msg) =>
        logger.error(msg)
        throw new RuntimeException()
      case Right((grade, msg)) =>
        // NOTE: do not set graded by person here, one who makes a record will get the honor
        exam.grade = grade
        exam.gradedTime = DateTime.now
        exam.creditType = exam.examType
        exam.examLanguages.asScala.headOption match
          case Some(lang) =>
            exam.answerLanguage = lang.code
            exam.state = ExamState.GRADED
            exam.update()
            logger.info(msg)
          case _ => throw new RuntimeException("No exam language found!")

  private def getGradeBasedOnScore(exam: Exam) =
    val (score, maxScore) = (exam.getTotalScore, exam.getMaxScore)
    val percentage        = if maxScore == 0 then 0 else score * 100 / maxScore
    val evaluations =
      exam.autoEvaluationConfig.gradeEvaluations.asScala.toList.sortBy(_.percentage)
    evaluations.findLast(_.percentage <= percentage) match
      case None => Left("Could not determine a grade")
      case Some(ge) =>
        resolveScale(exam) match
          case Some(s) if s.grades.contains(ge.grade) =>
            val grade = ge.grade
            val msg =
              s"Automatically grading exam #${exam.id}, $score/$maxScore points ($percentage%) " +
                s"graded as ${grade.name} using percentage threshold ${ge.percentage}"
            Right((grade, msg))
          case _ => Left("Grade in auto evaluation configuration not found in exam grade scale!")

  private def resolveScale(exam: Exam): Option[GradeScale] = Option(exam.gradeScale) match
    case scale @ Some(_) => scale
    case _ =>
      Option(exam.course).flatMap(c => Option(c.gradeScale)) match
        case scale @ Some(_) => scale
        case _ => Option(exam.parent).flatMap(p => Option(p.course)).flatMap(c =>
            Option(c.gradeScale)
          )
